src/ts/main/src/main.ts

```diff
@@ -0,0 +1,120 @@
+import {prepend} from "basic/prepend"
+import {PROJECT_PYTHON_FILES} from "./managed-by-build"
+import {CssScope, reduxLoopBasedImprovedLoop, ShadowRootContext} from "app/framework"
+import * as RootModule from "app/root"
+import {InvokeSubscriptionFunction} from "../../app/framework/src/subscriber"
+import {Action, createStore, Store} from "redux"
+import {install, StoreCreator} from "redux-loop"
+import {storageAction} from "app/storage"
+
+declare const languagePluginLoader: Promise<any>
+declare const pyodide: any
+
+function main() {
+    const start = (<any>window).start
+    console.log(new Date().getTime() - start)
+
+    console.log(prepend("hello ", "world, from TypeScript"))
+    console.log(PROJECT_PYTHON_FILES)
+
+    const enhancedCreateStore = createStore as StoreCreator
+    const store: Store<RootModule.RootState, Action> =
+        enhancedCreateStore(
+            RootModule.createReducer({
+                improvedLoop: reduxLoopBasedImprovedLoop,
+            }),
+            RootModule.initRootState(),
+            install())
+
+
+    document.addEventListener("DOMContentLoaded", function () {
+
+        document.body.ondragover = (e) => {
+            e.preventDefault()
+            e.stopPropagation()
+        }
+
+        document.body.ondrop = (e) => {
+            e.preventDefault()
+            e.stopPropagation()
+
+            const fileList = e.dataTransfer!.files
+            for (let i = 0; i < fileList.length; i++) {
+                const reader = new FileReader()
+                reader.addEventListener('load', (event) => {
+                    store.dispatch({
+                        type: storageAction.Keys.RECEIVE_DROPPED_FILE,
+                        name: fileList[i].name,
+                        content: (<string>event.target!.result)
+                    })
+                })
+                reader.readAsText(fileList[i])
+            }
+        }
+
+
+        const shadowRootContext = new ShadowRootContext(document.body.attachShadow({mode: "open"}), CssScope.ROOT)
+        shadowRootContext.render(RootModule.render())
+
+        RootModule.subscribe(
+            (subscriberFunction: InvokeSubscriptionFunction<RootModule.RootState>,
+             subscriberShadowContext: ShadowRootContext) => {
+                store.subscribe(() => {
+                    subscriberFunction(store.getState(), store.dispatch, shadowRootContext)
+                })
+            }, shadowRootContext)
+
+        // store.dispatch({
+        //     type: PageActionKeys.INIT
+        // })
+
+
+        languagePluginLoader.then(() => {
```

Comment:
what is this set to? Looks like it's undefined... must be somewhere else but I missed it.

Reply:
It's magic that comes from pyodide, note the `declare` statement at the top of `main.ts`